File: Data Structures & Algorithms/integer-break/submission-4.cpp
class Solution {
public:
    int integerBreak(int n) {
        unordered_map <int, int> dp;
        dp[1] = 1;
        for (int num = 1; num <= n; num++) {
            dp[num] = num == n ? 0 : num;
            for (int i = 1; i < num; i++) {
                int val = dp[i] * dp[num-i];
                dp[num] = max(dp[num], val);
            }
        }
        return dp[n];
    }
    
};

File: Data Structures & Algorithms/grumpy-bookstore-owner/submission-4.py
class Solution:
    def maxSatisfied(self, cust: List[int], grumpy: List[int], minutes: int) -> int:
        tot = 0
        l = [0] * len(cust)
        g = 0
        for i in range(len(grumpy)):
            if grumpy[i] == 0:
                tot+=cust[i]
            if grumpy[i] == 1:
                g+=cust[i]
            if i >= minutes and grumpy[i - minutes] == 1:
                g-=cust[i - minutes]
            l[i] = g
        return tot + max(l)